eth/downloader/queue.go

```diff
@@ -473,60 +757,94 @@ func (q *queue) Deliver61(id string, blocks []*types.Block) (err error) {
 		q.hashQueue.Push(hash, float32(index))
 	}
 	// If none of the blocks were good, it's a stale delivery
-	if len(errs) != 0 {
-		if len(errs) == len(blocks) {
-			return errStaleDelivery
-		}
+	switch {
+	case len(errs) == 0:
+		return nil
+
+	case len(errs) == 1 && (errs[0] == errInvalidChain || errs[0] == errInvalidBlock):
+		return errs[0]
+
+	case len(errs) == len(request.Headers):
```

Comment:
OK, would be great to put a short comment describing this above this case, as matching first on specific errors and then on length is a bit non-trivial.